src/everest/detached/jobs/everest_server_api.py

```diff
@@ -0,0 +1,354 @@
+    )
+
+    # Write certificate and key to disk
+    makedirs_if_needed(cert_folder)
+    cert_path = os.path.join(cert_folder, cert_name + ".crt")
+    with open(cert_path, "wb") as f:
+        f.write(cert.public_bytes(serialization.Encoding.PEM))
+    key_path = os.path.join(cert_folder, cert_name + ".key")
+    pw = bytes(os.urandom(28))
+    with open(key_path, "wb") as f:
+        f.write(
+            key.private_bytes(
+                encoding=serialization.Encoding.PEM,
+                format=serialization.PrivateFormat.TraditionalOpenSSL,
+                encryption_algorithm=serialization.BestAvailableEncryption(pw),
+            )
+        )
+    return cert_path, key_path, pw
+
+
+def _sim_monitor(context_status, event=None, shared_data=None):
+    status = context_status["status"]
+    assert shared_data
+    shared_data[SIM_PROGRESS_ENDPOINT] = {
+        "batch_number": context_status["batch_number"],
+        "status": {
+            "running": status.get("Running", 0),
+            "waiting": status.get("Waiting", 0),
+            "pending": status.get("Pending", 0),
+            "complete": status.get("Finished", 0),
+            "failed": status.get("Failed", 0),
+        },
+        "progress": context_status["progress"],
+    }
+
+    if shared_data[STOP_ENDPOINT]:
+        return "stop_queue"
+
+
+def _opt_monitor(shared_data=None):
+    assert shared_data
+    if shared_data[STOP_ENDPOINT]:
+        return "stop_optimization"
+
+
+class ExperimentRunnerStatus(BaseModel):
+    status: Optional[str] = None
+    exit_code: Optional[EverestRunModelExitCode | OptimizerExitCode] = None
+    message: Optional[str] = None
+
+
+class ExperimentRunner(threading.Thread):
+    def __init__(self, everest_config, state: dict):
+        super().__init__()
+
+        self.everest_config = everest_config
+        self.state = state
+        self.status: Optional[ExperimentRunnerStatus] = None
+
+    def run(self):
+        run_model = EverestRunModel.create(
+            self.everest_config,
+            simulation_callback=partial(_sim_monitor, shared_data=self.state),
+            optimization_callback=partial(_opt_monitor, shared_data=self.state),
+        )
+
+        evaluator_server_config = EvaluatorServerConfig(
+            custom_port_range=range(49152, 51819)
+            if run_model.ert_config.queue_config.queue_system == QueueSystem.LOCAL
+            else None
+        )
+
+        try:
+            run_model.run_experiment(evaluator_server_config)
+            self.status = ExperimentRunnerStatus(
+                status="Experiment finished", exit_code=run_model.exit_code
+            )
+        except Exception:
+            self.status = ExperimentRunnerStatus(
+                status="Experiment failed", message=traceback.format_exc()
+            )
+
+    def get_status(self) -> Optional[ExperimentRunnerStatus]:
+        return self.status
+
+
+security = HTTPBasic()
+
+
+class EverestServerAPI(threading.Thread):
```

Comment:
Looks like this code is mostly moved around? What is the advantage of using a class here instead of:
```python
def everserver_thread(output_dir: str, optimization_output_dir: str):
    app = FastAPI()
    state = {
        SIM_PROGRESS_ENDPOINT: {},
        STOP_ENDPOINT: False,
    }
    runner: Optional[ExperimentRunner] = None
    authentication = _generate_authentication()
    cert_path, key_path, key_pw = _generate_certificate(
        ServerConfig.get_certificate_dir(output_dir)
    )
    host = ert_shared_get_machine_name()
    port = _find_open_port(host, lower=5000, upper=5800)
    _write_hostfile(
        ServerConfig.get_hostfile_path(output_dir), host, port, cert_path, authentication
    )

    def check_user(credentials: HTTPBasicCredentials) -> None:
        if credentials.password != authentication:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials",
                headers={"WWW-Authenticate": "Basic"},
            )

    def log(request: Request) -> None:
        logging.getLogger("everserver").info(
            f"{request.scope['path']} entered from {request.client.host if request.client else 'unknown host'} with HTTP {request.method}"
        )

    @app.get("/")
    def get_status(request: Request, credentials: HTTPBasicCredentials = Depends(security)):
        log(request)
        check_user(credentials)
        ...
    ...
    
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=port,
        ssl_keyfile=key_path,
        ssl_certfile=cert_path,
        ssl_version=ssl.PROTOCOL_SSLv23,
        ssl_keyfile_password=key_pw,
        log_level=logging.CRITICAL,
    )
```

Personally I find it easier to read the endpoints when they are decorated, but there might be good reasons for making a class instead.

Reply:
Just to make the everserver.py file smaller and the code more readable

Reply:
I have no problem with moving it, but think the history between moving/refactoring and adding the new end-point, as it was a bit difficult to tell them apart. Personally I prefer the decorated functions instead of the class approach, as I find them easier to read.